Updated the code with the required changes to comply with ibmmq v2.0.3

tanaysingh3484 opened the pull request.

chughts commented on Node.js/README.md:
```diff
@@ -31,6 +31,12 @@ prerequsites by running :
 
 `npm install`
 
+Connecting to MQ with TLS
```
These instructions belong to the main repo readme. 

https://github.com/ibm-messaging/mq-dev-patterns/blob/master/README.md#tls

You can refer to that readme from here.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -174,7 +174,12 @@ class MQBoilerPlate {
 
       debug_info('Putting Message on Queue in mode ', mode);
-      mq.Put(queue, mqmd, pmo, msg, function(err) {
+      let putCall = mq.Put; 
```
Add comment as to why this is necessary.

chughts replied:
Needs rewording. As is it implies that the requester app is why the call needs to be sync, which is incorrect. 

The issue lies with the responder. This call is happening in the already async responder get callback. Adding another async on top of the async callback causes 2500. That is why this needs to be a sync put.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -257,15 +260,74 @@ class MQBoilerPlate {
-      mq.setTuningParameters( {
-        getLoopPollTimeMs: 500
-      });
+      // mq.setTuningParameters( {
```
Remove this code and the comments that go with it. Or replace with tuning parameters that can be set.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -257,15 +260,74 @@ class MQBoilerPlate {
     });
   }
 
+  startGetAsyncProcess() {
```
Add comment explaining what this function does.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -257,15 +260,74 @@ class MQBoilerPlate {
+    });
+  }
+
+  suspendAsyncProcess() {
```
Add comment explaining what this function does.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -257,15 +260,74 @@ class MQBoilerPlate {
+    });
+  }
+
+  resumeAsyncProcess() {
```
Add comment explaining what this function does.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -257,15 +260,74 @@ class MQBoilerPlate {
+  }
+
+
+  signalDone() {
```
Add comment explaining what this function does.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -257,15 +260,74 @@ class MQBoilerPlate {
+
   checkForTermination() {
     return new Promise(function resolver(resolve, reject) {
       var timerID = setInterval(() => {
```
could this var be converted into a let?

chughts commented on Node.js/boilerplate.js:
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright 2018, 2022 IBM Corp.
+ * Copyright 2018, 2024 IBM Corp.
```
I was expecting to see a lot more `var`s converted to `let's in this file.

chughts commented on Node.js/README.md:
```diff
@@ -1,6 +1,6 @@
 # IBM MQ Node.js samples
 The Node.js samples are based on https://github.com/ibm-messaging/mq-mqi-nodejs/tree/master/samples
-and have been tested with Node.js v16.8.1
+and have been tested with Node.js v20.11.1
```
Have the basicXXX set of examples been tested with this update. ie. basicget.js, basicput.js  etc.

tanaysingh3484 replied:
Yes I had run all the files in the repo when running the final regression test with the changes. All the changes made in the repo work with Node version 20.11.1

chughts replied:
I guess all the basicXX samples use synchronous mq operations, so they were immune to the ibmmq changes.

chughts commented on Node.js/samplerequest.js:
```diff
@@ -71,6 +71,10 @@ mqBoilerPlate.initialise('PUT')
     debug_info('Getting Response for ', msgID);
     return mqBoilerPlate.getMessagesDynamicQueue(msgID, msgCB);
   })
+  .then(() => {
```
Add comment explaining why this is necessary.

chughts commented on Node.js/sampleresponse.js:
```diff
@@ -61,13 +63,19 @@ async function msgCB(md, buf) {
-  
   if (!ok) {
-    mqBoilerPlate.rollback(buf, md, poisoningMessageHandler,callbackOnRollback);
+    mqBoilerPlate.suspendAsyncProcess()
```
Add comment explaining why this is necessary.

chughts commented on Node.js/sampleresponse.js:
```diff
@@ -101,13 +109,19 @@ function poisoningMessageHandler(buf,md) {
 function sendToQueue(buf, md, queue) {
-
   return mqBoilerPlate.openMQReplyToConnection(queue, 'DYNREP')
+  .then(() => {
```
Add comment explaining why this is necessary.

chughts replied:
This is incorrect. The suspend in in the response, and is on the async response get callback.

chughts replied:
No, the 1st async call isn't accessing the reply to queue, it is accessing the get queue. 

An async callback thread can't invoke an async MQ call.

chughts commented on Node.js/sampleresponse.js:
```diff
@@ -101,13 +109,19 @@ function poisoningMessageHandler(buf,md) {
+  })
+  .then(() => {
+    debug_info('Resuming the async get process');
+    return mqBoilerPlate.resumeAsyncProcess();
```
Add comment explaining why this is necessary.

chughts commented on Node.js/sampleresponse.js:
```diff
@@ -193,14 +207,23 @@ mqBoilerPlate.initialise('GET', true)
     debug_info('Getting Messages');
     return mqBoilerPlate.getMessages(null, msgCB);
   })
+  .then(() => {
```
Add comment.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -531,11 +601,17 @@ class MQBoilerPlate {
-        mq.Close(hObjSubscription, 0);
-        debug_info("MQCLOSE (Subscription) successful");
+        mq.Close(hObjSubscription, 0,function (err){
+          if(err){
```
Need to conform to house styling of spaces around if block.

tanaysingh3484 commented:
Have made the changes.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -254,21 +263,78 @@ class MQBoilerPlate {
+    debug_info('Resuming callback');
+    let me = this;
+    return new Promise(function resolver(resolve, reject){
+      mq.Ctl(me.mqConn, MQC.MQOP_SUSPEND, (err) => {
```
Why is it still a suspend? It needs to be a RESUME. 

The function is `resumeAsyncProcess()` So the expectation is that it performs a resume.

tanaysingh3484 commented:
> Why is it still a suspend? It needs to be a RESUME. The function is resumeAsyncProcess() So the expectation is that it performs a resume.

Adding a MQC.MQOP_RESUME, generated the following error in the Response file: 

```
samplerep:warn Error on commit MQError: CMIT: MQCC = MQCC_FAILED [2] MQRC = MQRC_HCONN_ASYNC_ACTIVE [2500]
    at exports.Cmit (C:\Users\TanaySingh\Documents\mq-dev-patterns\Node.js\node_modules\ibmmq\lib\mqi.js:921:13)
    at MQBoilerPlate.commit (C:\Users\TanaySingh\Documents\mq-dev-patterns\Node.js\boilerplate.js:219:8)
    at handleSyncPoint (C:\Users\TanaySingh\Documents\mq-dev-patterns\Node.js\sampleresponse.js:81:19)
    at msgCB (C:\Users\TanaySingh\Documents\mq-dev-patterns\Node.js\sampleresponse.js:57:5) {
  mqcc: 2,
  mqccstr: 'MQCC_FAILED',
  mqrc: 2500,
  mqrcstr: 'MQRC_HCONN_ASYNC_ACTIVE',
  version: '2.0.3',
  verb: 'CMIT'
} +0ms
```

tanaysingh3484 commented:
New Commit consists of the following changes : 
- Updated the comments for the `suspendAsyncProcess` and the `resumeAsyncProcess` functions. Also updated the comments for the `putCall` switch between `mq.Put` and `mq.PutSync`.
- Updated the `commit` and `rollback` functions in the `sampleresponse` file, to shift the callbacks present in them to be anonymous functions in the `boilerplate` and correctly resolve or reject the promises returned from `mq.Back` and `mq.Cmit`.
- Since the app was not dying gracefully, updated the `samplerequest` and the `sampleresponse` files to terminate in the `.then()` chain rather than after the Promise chain.
Thanks

chughts commented on Node.js/sampleresponse.js:
```diff
@@ -40,14 +40,16 @@ var mqBoilerPlate = new MQBoilerPlate();
 
 async function msgCB(md, buf) {
```
msgCB is an async process : which means that it is returning a promise. That means that the callback passed into `mqBoilerPlate.getMessages` could be receiving either a normal function or a function that returns a promise. 

That means that in boilerplate:

```
  getCallback(err, hObj, gmo, md, buf) {
    // If there is an error, prepare to exit by setting the ok flag to false.
    if (err) {
      if (err.mqrc == MQC.MQRC_NO_MSG_AVAILABLE) {
        debug_info("No more messages available.");
      } else {
        MQBoilerPlate.reportError(err);
      }
      canExit = true;
      // We don't need any more messages delivered, so cause the
      // callback to be deleted after this one has completed.
      //mq.GetDone(hObj);
    } else {
      if (activeCB) {
        canExit = !activeCB(md, buf);
      }
    }
  }
  ```

The lines
```
      if (activeCB) {
        canExit = !activeCB(md, buf);
      }
      ```
      
      Needs to check if the return is a promise or a value. ie. 
      
      ```
      if (activeCB) {
        if (activeCB instanceof Promise) {
           activeCB(md, buf)
           .then((mustContinue) => {canExit = ! mustContinue;})
        } else {
           canExit = !activeCB(md, buf);
        }
      }
      ```

This might fix the need to use process.exit()

chughts commented on Node.js/sampleresponse.js:
```diff
@@ -40,14 +40,16 @@ var mqBoilerPlate = new MQBoilerPlate();
+      if (ok) {
+        debug_info('Starting response sequence');
+        ok = await respondToRequest(msgObject, md);
+      }
```
This function is always returning true. ie. the async promise is always resolving to true. It needs to return `ok`, so that its promise can resolve to false.

chughts commented on Node.js/sampleresponse.js:
```diff
@@ -61,13 +63,25 @@ async function msgCB(md, buf) {
-    mqBoilerPlate.commit(callbackOnCommit);
-  }
-
+  //Suspending the background async get process to avoid MQRC 2500 : MQRC_HCONN_ASYNC_ACTIVE.
```
handleSyncPoint invokes a series of promises. So it should also be either marked as async or return a promise. Which in turn affects how it is invoked in `msgCB` ie. 
```
await handleSyncPoint(buf, md, ok);
```

tanaysingh3484 commented:
Added the following updates : 
- Updated the required comments
- Modified the `sampleresponse` and `samplerequest` files, with the following updates : 
   - Updated the exit in the `samplerequest` and `sampleresponse` to await the `teardown()` function, and then exit with `process.exit`
   - Updated the `msgCB` function in `sampleresponse` to return the value of `ok` rather than the previous `true` value.
   - Updated the `handleSyncPoint` function to be an async function, so that it can be awaited in the `msgCB` function.

- Updated the `boilerplate` file with changes in the `rollback` and `commit` functions to return promises.
- Updated the `teardown` function in the `boilerplate` to return a promise.
- Made some minor changes such as changing a `console.log` to `debug_info`, changing the poisition of the `debug_info('Starting Application')` in the `samplerequest` file.

chughts commented on Node.js/samplerequest.js:
```diff
@@ -71,15 +71,21 @@ mqBoilerPlate.initialise('PUT')
   })
   .then(() => {
     mqBoilerPlate.teardown();
+    debug_info('Application Completed');
```
Like this, this async anonymous function returns a promise. It might be better to return the teardown, and have a subsequent then that runs the process.exit if needed.

chughts commented on Node.js/samplerequest.js:
```diff
@@ -71,15 +71,22 @@ mqBoilerPlate.initialise('PUT')
   })
-  .catch((err) => {
-    mqBoilerPlate.teardown();
+  .catch(async (err) => {
```
Like this, this anonymous async function returns a promise.

chughts commented on Node.js/sampleresponse.js:
```diff
@@ -193,16 +207,27 @@ mqBoilerPlate.initialise('GET', true)
+    debug_info('Signal termination of the callback thread');
+    return mqBoilerPlate.signalDone();
+  })    
+  .then(async () => {
```
like this, this anonymous async function returns a promise.

tanaysingh3484 commented on Node.js/boilerplate.js:
```diff
@@ -595,13 +711,20 @@ class MQBoilerPlate {
       //mq.GetDone(hObj);
```
This `GetDone` was already present in the boilerplate when I had cloned the repo. I have not modified it, and left it commented out as it was. Do you want me to incorporate this into the code, or leave it as it is?

tanaysingh3484 commented:
In this commit, 
- Updated the `closeSubscription` function to return a Promise.
- Cleaned up the code in places where a Promise was being returned, for e.g. the `if (! hObj)` or the `if (! hConn)`.
- Updated the `.then` and `.catch` chains in `samplerequest` and `sampleresponse` to have an additional `.then` to wait for the  return of `teardown`.

chughts reviewed (changes requested):
Make the requested changes to refactor the 2 mid function promises into a separate private function that gets invoked.

chughts commented on Node.js/boilerplate.js:
```diff
@@ -595,13 +711,20 @@ class MQBoilerPlate {
       if (activeCB) {
-        canExit = !activeCB(md, buf);
+        if (activeCB instanceof Promise) {
+          activeCB(md,buf)
```
This debug message needs to provide some context. 

Something along the lines of "activeCB is a promise, so need to wait for it to resolve before determining if the getCallback handling can be stopped."

tanaysingh3484 commented:
In this commit, the following updates were made : 
- The code was cleaned up by removing the multiple suspends and resumes in the `sampleresponse` file, and shifting all of them into the `getCallback` function of the `boilerplate`.
- Two helper functions were added to further improve readability to check when the `activeCB` returned an instance of a `Promise` and when it didn't. 
- Updated the `sampleput` ,`sampleget`, `samplepublish` and `samplesubscribe` files to make sure the app dies gracefully.
- Added comments where necessary.
- Cleaned up the code by maintaining an uniform format throughout.

chughts commented on Node.js/samplesubscribe.js:
```diff
@@ -60,16 +60,30 @@ mqBoilerPlate.initialise('SUBSCRIBE')
-
-
-debug_info('Application Completed');
+  .then(() => {
```
This final then won't be called. If there is an error, then this .then needs to be on the teardown in the catch. ie.

mqBoilerPlate.teardown().then(() => process.exit(1))

tanaysingh3484 commented:
I looked into the other files as well for the same issue, and found out that the other `samples` also had this `.then()` after the `.catch()` , so I went ahead and updated those files as well.

chughts reviewed (commented):
approved, but you will need to create a fresh pull-request.
